libr/cons/dietline.c

```diff
@@ -1196,12 +1256,364 @@ R_API const char *r_line_readline_cb_win(RLineReadCallback cb, void *user) {
+			while (IS_DIGIT (c) && ch != '0') {
+				int tmp = c - '0';
+				rep1 = (rep1 * 10) + tmp;
+				c = r_cons_readchar ();
+			}
+			rep1 = rep1 > 0? rep1: 1;
+			while (rep1--) {
+				switch (c) {
+				case 'i': {
+					char t = r_cons_readchar ();
+					if (t == 'w') {
+						kill_word ();
+					} else if (t == 'b') {
+						backward_kill_word ();
+					} else if (t == 'W') {
+						kill_Word ();
+					} else if (t == 'B') {
+						backward_kill_Word ();
+					}
+				} break;
+				case 'W':
+					kill_Word ();
+					break;
+				case 'w':
+					kill_word ();
+					break;
+				case 'B': 
+					backward_kill_Word ();
+					break;
+				case 'b':
+					backward_kill_word ();
+					break;
+				case 'h':
+					delete_prev_char ();
+					break;
+				case 'l':
+					delete_next_char ();
+					break;
+				case '$':
+					delete_till_end ();
+					break;
+				case '^':
+				case '0':
+					strncpy (I.buffer.data, I.buffer.data + I.buffer.index, I.buffer.length);
+					I.buffer.length -= I.buffer.index;
+					I.buffer.index = 0;
+					break;
+				} __print_prompt ();
+			}
+			} break;
+		case 'I':
+			mode = INSERT_MODE;
+		case '^':
+		case '0': 
+			if (gcomp) {
+				strcpy (I.buffer.data, gcomp_line);
+				I.buffer.length = strlen (I.buffer.data);
+				I.buffer.index = 0;
+				gcomp = false;
+			}
+			I.buffer.index = 0;
+			break;
+		case 'A':
+			mode = INSERT_MODE;
+		case '$': 
+			if (gcomp) {
+				strcpy (I.buffer.data, gcomp_line);
+				I.buffer.index = strlen (I.buffer.data);
+				I.buffer.length = I.buffer.index;
+				gcomp = false;
+			} else {
+				I.buffer.index = I.buffer.length;
+			} break;
+		case 'p': 
+			while (rep--) {
+				paste ();
+				__print_prompt();
+			} break;
+		case 'a':
+#if USE_UTF8
+			{
+				char *s = I.buffer.data + I.buffer.index - 1;
+				utflen = 1;
+				while (s > I.buffer.data && (*s & 0xc0) == 0x80) {
+					utflen++;
+					s--;
+				}
+			}
+			I.buffer.index = I.buffer.index
+					 ? I.buffer.index + utflen
+					 : 0;
+#else
+			I.buffer.index = I.buffer.index
+					 ? I.buffer.index + 1
+					 : 0;
+#endif
+			__print_prompt ();
+		case 'i': {
+			mode = INSERT_MODE;
+			break;
+		}
+		case 75:
+		case 'h': 
+			while (rep--) {
+				__move_cursor_left ();
+				__print_prompt ();
+			} break;
+		case 77:
+		case 'l': 
+			while (rep--) {
+				__move_cursor_right ();
+				__print_prompt ();
+			} break;
+		case 'E':
+			while (rep--) {
+				for (i = I.buffer.index + 1; i < I.buffer.length; i++) {
+					if ((!is_word_break_char (I.buffer.data[i], MAJOR_BREAK) && is_word_break_char (I.buffer.data[i+1], MAJOR_BREAK))) {
+						I.buffer.index = i;
+						break;
+					}
+				}
+				if (i >= I.buffer.length) {
+					I.buffer.index = I.buffer.length - 1;
+				}
+				__print_prompt();
+		} break;
+		case 'e':
+			while (rep--) {
+				for (i = I.buffer.index + 1; i < I.buffer.length; i++) {
+					if ((!is_word_break_char (I.buffer.data[i], MINOR_BREAK) && is_word_break_char (I.buffer.data[i+1], MINOR_BREAK))
```

Comment:
What about separating these into small, maybe inline functions?

Reply:
Updated the code :)